ctapipe/io/eventfilereader.py

```diff
@@ -149,215 +98,202 @@ def check_file_compatibility(file_path):
+        waveforms, as some cameras require different algorithms to others
+        (such as with R1 calibration and charge extraction).
 
         Returns
         -------
-        self._num_events : int
+        camera_name : str
         """
 
     @property
-    @abstractmethod
-    def event_id_list(self):
+    def is_stream(self):
         """
-        Abstract property to be defined in child class.
+        Bool indicating if input is a stream. If it is then `__getitem__` is
+        disabled.
 
-        Obtain the number of events from the file, store it as
-        self._event_id_list, and return the value.
+        TODO: Define a method to detect if it is a stream
 
         Returns
         -------
-        self._event_id_list : list[self._num_events]
+        bool
+            If True, then input is a stream.
         """
+        return False
 
     @abstractmethod
-    def read(self, allowed_tels=None, requested_event=None,
-             use_event_id=False):
+    def _generator(self):
         """
         Abstract method to be defined in child class.
 
-        Read the file using the processes required by the readers file-type.
-
-        Parameters
-        ----------
-        allowed_tels : list[int]
-            select only a subset of telescope, if None, all are read. This can
-            be used for example emulate the final CTA data format, where there
-            would be 1 telescope per file (whereas in current monte-carlo,
-            they are all interleaved into one file)
-        requested_event : int
-            Seek to a paricular event index
-        use_event_id : bool
-            If True ,'requested_event' now seeks for a particular event id
-            instead of index
+        Generator where the filling of the `ctapipe.io.containers` occurs.
 
         Returns
         -------
-        source : generator
-            A generator that can be iterated over to obtain events
+        generator
         """
 
-    def get_event(self, requested_event, use_event_id=False):
+    def reset(self):
         """
-        Loop through events until the requested event is found
-
-        Parameters
-        ----------
-        requested_event : int
-            Seek to a paricular event index
-        use_event_id : bool
-            If True ,'requested_event' now seeks for a particular event id
-            instead of index
-
-        Returns
-        -------
-        event : `ctapipe` event-container
-
+        Recreate the generator so it starts from the beginning
         """
-        source = self.read(requested_event=requested_event,
-                           use_event_id=use_event_id)
-        event = next(source)
-        return deepcopy(event)
+        self.source = self._generator()
+        self.current_event = None
 
-    def find_max_true_npe(self, telescopes=None):
+    def __iter__(self):
+        # Always reset generator when starting a new iteration
+        self.reset()
```

Comment:
I think this is very counter intuitive. When I do:

```
reader = EventFileReader()
event = next(reader)
for event in reader:
    pass
```
Would this start with event 0 in the loop again?

Reply:
I guess that's the difference between a generator and an iterable. The latter would re-start and the former would continue... If we want these to be generator-like, it makes sense to have the user explicitly do a reset, but that can also be confusing for people not familiar with generators.

Reply:
Firstly there is no `__next__` method defined. Again, EventFileReader is currently an [iterable](https://docs.python.org/3/glossary.html#term-iterable) not an [iterator](https://docs.python.org/3/glossary.html#term-iterator) .

But yes, every time you start a new `for event in reader:` it starts from event 0 again. 
If you didn't do that an operation like:

```
for event in reader:
   if event.count == 5: 
       break
for event in reader:
   if event.count == 2: 
       break
```

Wouldn't work. For me its more intuitive that the loop starts at the beginning of the file again when starting a new loop. Do you disagree?

Reply:
When you do:

```
example_list = [3, 4, 6, 8, 9]
for el in example_list:
   if el == 8:
      break
for el in example_list:
   if el == 4:
      break
```

You expect it to work...

Reply:
I think it's fine this way as long as it's documented.  In fact, there have been a few people confused that `hessio_event_source` didn't restart this way (since most people new to Python are not really familiar with generators)